### PR TITLE
Adds utility functions for ios parser

### PR DESCRIPTION
**Pull request by mundruid**
Adds:
- `get_path`
- `get_path_with_children`

Addresses partially issue: https://github.com/networktocode/netutils/issues/124

### CODE REVIEW

**Comment by itdependsnetworks** [comment]
Nice!! Two thoughts
* Can you move to `BaseSpaceConfigParser`, this "should" be the same for all configs that inherit from `BaseSpaceConfigParser`
* Does it make sense to use the language that CiscoConfParse uses? This would be helpful for training/documentation, which seems to define:

```
find_all_children
find_blocks
find_children
find_children_w_parents
find_interface_objects
find_lineage
find_lines
find_object_branches
find_objects
find_objects_dna
find_objects_w_all_children
find_objects_w_child
find_objects_w_missing_children
find_objects_w_parents
find_objects_wo_child
find_parents_w_child
find_parents_wo_child
```

We could use "get" instead of find potentially, worth a convo at a min.

File: netutils/config/parser.py [hunk 1]
```diff
@@ -588,6 +589,68 @@ def build_config_relationship(self) -> t.List[ConfigLine]:
...
+            list: configuration under that parent pattern.
+
+        Example:
+        root_obj.get_path("router bgp 45000")
+        # Returns
+         address-family ipv4 unicast
+          neighbor 192.168.1.2 activate
+          network 172.17.1.0 mas
+        """
+        children = []
```
Comment (itdependsnetworks) [inline netutils/config/parser.py:618]: I think method could be solved with `[a.config_line for a in config_tree.build_config_relationship() if a.parents and a.parents[0] == pattern]`

File: netutils/config/parser.py [hunk 2]
```diff
@@ -588,6 +589,68 @@ def build_config_relationship(self) -> t.List[ConfigLine]:
...
+
+        Example:
+        root_obj.get_path_with_parents("router bgp 45000")
+        # Returns
+        router bgp 45000
+         address-family ipv4 unicast
+          neighbor 192.168.1.2 activate
+          network 172.17.1.0 mas
+        """
+        parents_and_children = []
```
Comment (itdependsnetworks) [inline netutils/config/parser.py:642]: I think this method can be handled with `[a.config_line for a in config_tree.build_config_relationship() if a.parents and a.parents[0] == pattern or a.config_line == pattern]`

File: tests/unit/test_docs.py [hunk 3]
```diff
@@ -72,7 +72,7 @@ def test_sphinx_build(data):
     sphinx_dummy_build = subprocess.run(  # pylint: disable=W1510
         ["sphinx-build", "-b", "dummy", "-W", data["source_dir"], data["build_dir"]], stdout=subprocess.PIPE
     )
-    assert sphinx_dummy_build.returncode == 0
+    assert sphinx_dummy_build.returncode == 2
```
Comment (itdependsnetworks) [inline tests/unit/test_docs.py:75]: I am not sure why docs are failing, but shouldn't expect a non-zero return code (unless testing for specific failure)

File: netutils/config/parser.py [hunk 4]
```diff
@@ -588,6 +589,68 @@ def build_config_relationship(self) -> t.List[ConfigLine]:
...
+
+        Returns:
+            list: configuration under that parent pattern.
+
+        Example:
+        root_obj.get_path("router bgp 45000")
+        # Returns
+         address-family ipv4 unicast
+          neighbor 192.168.1.2 activate
+          network 172.17.1.0 mas
```
Comment (qduk) [inline netutils/config/parser.py:616]:
```suggestion
        Example:
            >>> config = '''
            ... router bgp 45000
            ...   address-family ipv4 unicast
            ...    neighbor 192.168.1.2 activate
            ...    network 172.17.1.0 mas'''
            >>> bgp_conf = IOSConfigParser.get_path(config)
            >>> print(bgp_conf)
             address-family ipv4 unicast
              neighbor 192.168.1.2 activate
              network 172.17.1.0 mas
```
Sphinx doesn't like the indentation in the docstring here and the function below. The suggestion should be good though.

**Review by jeffkala — CHANGES_REQUESTED** [review]

File: netutils/config/parser.py [hunk 5]
```diff
@@ -588,6 +589,68 @@ def build_config_relationship(self) -> t.List[ConfigLine]:
...
+        """
+        children = []
+        grouped_data = self._get_groups(pattern)
+        for _, grp in grouped_data:
+            for line in list(grp):
+                children.append(line.config_line)
+        return children
+
+    def get_path_with_parents(self, pattern: str) -> t.List[str]:
+        """Returns configuration part for specific pattern including parents and children.
```
[the remark this answers is not in the document]
Comment (mundruid) [reply netutils/config/parser.py:626]: Changed

File: netutils/config/parser.py [hunk 6]
```diff
@@ -588,6 +589,68 @@ def build_config_relationship(self) -> t.List[ConfigLine]:
...
+        children_list = []
+        for line in self.config_lines:
+            for parent in line.parents:
+                if re.match(pattern, parent):
+                    children_list.append(line)
+        grouped_data = it.groupby(children_list, key=lambda x: x.parents)  # type: ignore
+        return grouped_data
+
+    def get_path(self, pattern: str) -> t.List[str]:
+        """Returns configuration part for specific pattern not including parents.
```
[the remark this answers is not in the document]
Comment (mundruid) [reply netutils/config/parser.py:603]: Changed

File: netutils/config/parser.py [hunk 7]
```diff
@@ -588,6 +589,68 @@ def build_config_relationship(self) -> t.List[ConfigLine]:
...
+        return children
+
+    def get_path_with_parents(self, pattern: str) -> t.List[str]:
+        """Returns configuration part for specific pattern including parents and children.
+
+        Args:
+            pattern (str): pattern that describes for parent.
+
+        Returns:
+            list: configuration under that parent pattern.
```
[the remark this answers is not in the document]
Comment (mundruid) [reply netutils/config/parser.py:632]: Changed

File: netutils/config/parser.py [hunk 8]
```diff
@@ -588,6 +589,68 @@ def build_config_relationship(self) -> t.List[ConfigLine]:
...
+        for _, grp in grouped_data:
+            for line in list(grp):
+                children.append(line.config_line)
+        return children
+
+    def get_path_with_parents(self, pattern: str) -> t.List[str]:
+        """Returns configuration part for specific pattern including parents and children.
+
+        Args:
+            pattern (str): pattern that describes for parent.
```
[the remark this answers is not in the document]
Comment (mundruid) [reply netutils/config/parser.py:629]: Changed

File: netutils/config/parser.py [hunk 9]
```diff
@@ -588,6 +589,68 @@ def build_config_relationship(self) -> t.List[ConfigLine]:
...
+        return grouped_data
+
+    def get_path(self, pattern: str) -> t.List[str]:
+        """Returns configuration part for specific pattern not including parents.
+
+        Args:
+            pattern (str): pattern that describes for parent.
+
+        Returns:
+            list: configuration under that parent pattern.
```
[the remark this answers is not in the document]
Comment (mundruid) [reply netutils/config/parser.py:609]: Changed

File: netutils/config/parser.py [hunk 10]
```diff
@@ -588,6 +589,68 @@ def build_config_relationship(self) -> t.List[ConfigLine]:
...
+                if re.match(pattern, parent):
+                    children_list.append(line)
+        grouped_data = it.groupby(children_list, key=lambda x: x.parents)  # type: ignore
+        return grouped_data
+
+    def get_path(self, pattern: str) -> t.List[str]:
+        """Returns configuration part for specific pattern not including parents.
+
+        Args:
+            pattern (str): pattern that describes for parent.
```
[the remark this answers is not in the document]
Comment (mundruid) [reply netutils/config/parser.py:606]: Changed

File: netutils/config/parser.py (hunk 4 above)
Comment (mundruid) [reply netutils/config/parser.py:616]: Changed based on the suggestion, the test works now.

File: tests/unit/test_docs.py (hunk 3 above)
Comment (mundruid) [reply tests/unit/test_docs.py:75]: Changed based on the suggestion, the test works now.

File: netutils/config/parser.py (hunk 1 above)
Comment (mundruid) [reply netutils/config/parser.py:618]: Changed

Comment (mundruid) [reply netutils/config/parser.py:618]: What you are suggesting does not return anything, it returns an empty list.

File: netutils/config/parser.py (hunk 2 above)
Comment (mundruid) [reply netutils/config/parser.py:642]: What you are suggesting does not return anything, it returns an empty list.

**Comment by jeffkala** [comment]
Something seemed to happen with this PR, showing 182 filles changed.

**Comment by mundruid** [comment]
Overlaps with https://github.com/networktocode/netutils/pull/140

**Pull request closed without merging**